storage-nvmf/src/main/java/org/apache/crail/storage/nvmf/client/NvmfStorageEndpoint.java

```diff
@@ -19,208 +19,229 @@
+			if (op == Operation.READ) {
+				throw new IOException("Unaligned read access is not supported. Address (" + startingAddress +
+						") needs to be multiple of LBA data size " + getLBADataSize());
+			}
 			try {
-				switch(op) {
-					case READ: {
-						NvmfStorageFuture f = futures[(int)command.getId()] = new NvmfStorageFuture(this, (int)alignedLength);
-						command.setBuffer(stagingBuffer.getByteBuffer()).setLinearBlockAddress(lba).read().execute();
-						future = new NvmfStorageUnalignedReadFuture(f, this, buffer, remoteMr, remoteOffset, stagingBuffer);
-						break;
-					}
-					case WRITE: {
-						if (NvmfStorageUtils.namespaceSectorOffset(sectorSize, remoteOffset) == 0) {
-							// Do not read if the offset is aligned to sector size
-							int sizeToWrite = length;
-							stagingBuffer.put(buffer.getByteBuffer());
-							stagingBuffer.position(0);
-							command.setBuffer(stagingBuffer.getByteBuffer()).setLinearBlockAddress(lba).write().execute();
-							future = futures[(int)command.getId()] = new NvmfStorageUnalignedWriteFuture(this, sizeToWrite, stagingBuffer);
-						} else {
-							// RMW but append only file system allows only reading last sector
-							// and dir entries are sector aligned
-							stagingBuffer.limit(sectorSize);
-							NvmfStorageFuture f = futures[(int)command.getId()] = new NvmfStorageFuture(this, sectorSize);
-							command.setBuffer(stagingBuffer.getByteBuffer()).setLinearBlockAddress(lba).read().execute();
-							future = new NvmfStorageUnalignedRMWFuture(f, this, buffer, remoteMr, remoteOffset, stagingBuffer);
-						}
-						break;
-					}
-				}
-			} catch (NoSuchFieldException e) {
-				throw new IOException(e);
-			} catch (IllegalAccessException e) {
+				return new NvmfUnalignedWriteFuture(this, buffer, blockInfo, remoteOffset);
+			} catch (Exception e) {
 				throw new IOException(e);
 			}
 		}
 
+		if (!tryGetOperation()) {
+			do {
+				poll();
+			} while (!tryGetOperation());
+		}
+
+		NvmIoCommand<? extends NvmIoCommandCapsule> command;
+		NvmfFuture<?> future;
+		Response<NvmResponseCapsule> response;
+		if (op == Operation.READ) {
+			NvmReadCommand readCommand = readCommands.remove();
+			response = readCommand.newResponse();
+			future = new NvmfFuture<>(this, readCommand, response, readCommands, buffer.remaining());
+			command = readCommand;
+		} else {
+			NvmWriteCommand writeCommand = writeCommands.remove();
+			response = writeCommand.newResponse();
+			future = new NvmfFuture<>(this, writeCommand, response, writeCommands, buffer.remaining());
+			command = writeCommand;
+		}
+		command.setCallback(future);
+		response.setCallback(future);
+
+		NvmIoCommandSqe sqe = command.getCommandCapsule().getSubmissionQueueEntry();
+		long startingLBA = startingAddress / getLBADataSize();
+		sqe.setStartingLba(startingLBA);
+		/* TODO: on read this potentially overwrites data beyond the set limit */
+		short numLogicalBlocks = (short)getNumLogicalBlocks(buffer);
```

Comment:
Any particular reason this is short? than an integer?

Reply:
Not anymore. Previously the lower level API used shorts (since it is the max supported size). Will change.